Add DefaultQueryFilters

NiseVoid opened the pull request:
# Objective

Some usecases in the ecosystems are blocked by the inability to stop bevy internals and third party plugins from touching their entities. However the specifics of a general purpose entity disabling system are controversial and further complicated by hierarchies. We can partially unblock these usecases with an opt-in approach: default query filters.

## Solution

- Introduce DefaultQueryFilters, these filters are automatically applied to queries that don't otherwise mention the filtered component.
- End users and third party plugins can register default filters and are responsible for handling entities they have hidden this way.
- Extra features can be left for after user feedback
- The default value could later include official ways to hide entities

---

## Changelog

- Add DefaultQueryFilters

SkiFire13 commented on crates/bevy_ecs/src/query/state.rs:
```diff
@@ -2030,4 +2045,32 @@ mod tests {
+        let b = world.component_id::<B>().unwrap();
+        let c = world.component_id::<C>().unwrap();
+
+        let mut default_filters = DefaultQueryFilters::default();
+        default_filters.with(b);
+        default_filters.without(c);
+        world.insert_resource(default_filters);
+
+        // With<B>, Without<C> only matches the first entity
+        let mut query = QueryState::<()>::new(&mut world);
+        assert_eq!(1, query.iter(&world).count());
```
I worry about what would happen if the order of setting up `DefaultQueryFilters` and the `QueryState` is inverted. Would the `QueryState` ignore the filters? That feels like a huge footgun.

SkiFire13 replied:
This doesn't solve the issue when a plugin creates a `QueryState`/`SystemState` while building the plugin, as it won't use the "full"/"final" `DefaultQueryFilters`. Note that such plugin may not even use query filters itself, so the author might not have read the documentation of `DefaultQueryFilters`, so it would have to be mentioned in the documentation of `QueryState`/`SystemState`.

---

Also, since adding default filters after plugin construction is basically guaranteed to produce incorrect behaviour I would propose to make that panic if it happens. Better failing fast than breaking things :)

The caveats should also be mentioned in `World::set_default_with_filter` and `World::set_default_without_filter` since those are the methods that people will use most often. Or at least mention the fact that there are caveats and to look at the documentation for `DefaultQueryFilters` to see them (currently it links to it but doesn't mention the fact there are caveats).

iiYese commented on crates/bevy_ecs/src/query/default_filters.rs:
```diff
@@ -0,0 +1,167 @@
+/// These filters are only applied to queries initialized after updating this resource,
+/// it should most likely only be modified before the app starts.
+#[derive(Resource, Default)]
+pub struct DefaultQueryFilters(Vec<DefaultFilter>);
```
Maybe insert it by default so plugin authors aren't needing to check `Option`s all the time to avoid overwriting their default filters.

NiseVoid replied:
Where do we want to insert it? Looks like we have nothing fitting in `bevy_core` ... But we could add a new plugin there ... Or it could be added to `App:default()`

iiYese replied:
App works since it's just this atm unless someone else knows somewhere better.

iiYese replied:
Might want to move it into its own plugin if we decide to add a bunch of default filters (like excluding window & systems).

iiYese commented:
Two things you should probably do:
- Add a way to bypass default filters.
- Change `.retain`. Make it so it doesn't remove components that are filtered as `Without` & add another variant that bypasses that.

NiseVoid commented:
> Add a way to bypass default filters.

This seems reasonable, yet it might also be problematic for things that get added to the list in the future. It's hard to anticipiate all the uses for this before people have a chance to use it. It's hard to say if a universal bypass makes sense or if we'd be better off with some alternative approach like assigning a group to each filter, and allowing groups to be bypassed. Or perhaps bypassing default filters won't even be needed in practice. It might be better to leave it until people report running into issues, or until we start including default features and know we need a way to bypass them collectively.

> Change `.retain`. Make it so it doesn't remove components that are filtered as `Without` & add another variant that bypasses that.

This behavior feels a bit implicit, but I think creating a list of components retain shouldn't touch could make a lot of sense, it would also be a lot easier to explain and reason about for users. There might even be uses that do want to bypass retain but not be filtered out, or the other way around. But it should probably be in a separate PR. The bypass variant might also have the same issues as the point above.

iiYese commented:
> This behavior feels a bit implicit, but I think creating a list of components retain shouldn't touch could make a lot of sense

At the very least it should be the same list. It shouldn't be a different one. I still think it would be a good idea to include this now anyway even if you don't want to add bypasses yet because introducing it later will be a breaking change & using it without default filters applied will surprise users.

NiseVoid commented:
> At the very least it should be the same list.

I'm thinking more along the lines: Most users wouldn't manually add default filters, we'll probably want built-in abstractions on top in a future release. Meaning they'd probably be configured by plugins (something like a `DisabledPlugin<T>` or a `HidePlugin<T>`), and at this point the list of plugins serves as the list, which this list and a "retain filters" list could then be derived from.

Tho I've never actually used retain so it's a bit hard for me to comment on how it should behave.

Shatur reviewed (approved):
The code is simple and clear. I would add references to `DefaultQueryFilters` from `Query` for discoverability.

This feature would be quite useful for rollback in networking.

cBournhonesque commented on crates/bevy_ecs/src/query/default_filters.rs:
```diff
@@ -0,0 +1,167 @@
+    kind: FilterKind,
+}
+
+/// A list of default query filters, these can be used to globally exclude entities from queries.
```
I don't think it would be completely obvious to me what this means without an example, but it's not blocking the PR.

A little example inside the docstring with a `Disabled` filter could be useful, to explain that this can be used to affect queries outside of your crate.

cBournhonesque reviewed (approved):
Seems well thought-out, and definitely useful

NiseVoid commented on crates/bevy_ecs/src/query/default_filters.rs:
```diff
@@ -0,0 +1,167 @@
+    }
+}
+
+fn contains_component(
```
The way it functions might not be the expected behavior of future logic needing something like it. But if other code needs the same check later it can of course be moved there.

alice-i-cecile commented on crates/bevy_ecs/src/query/default_filters.rs:
```diff
@@ -0,0 +1,173 @@
+///
+/// These filters are only applied to queries initialized after updating this resource,
+/// it should most likely only be modified before the app starts.
+#[derive(Resource, Default)]
```
This wants a Debug (and possibly Reflect) derive.

alice-i-cecile commented on crates/bevy_ecs/src/query/default_filters.rs:
```diff
@@ -0,0 +1,173 @@
+/// These filters are only applied to queries initialized after updating this resource,
+/// it should most likely only be modified before the app starts.
+#[derive(Resource, Default)]
+pub struct DefaultQueryFilters(Vec<DefaultFilter>);
```
Why is this a `Vec` rather than a `HashMap<ComponentId, FilterKind>`?

alice-i-cecile commented on crates/bevy_ecs/src/query/default_filters.rs:
```diff
@@ -0,0 +1,173 @@
+}
+
+/// A list of default query filters, these can be used to globally exclude entities from queries.
+/// Each individual filter is only applied to queries that don't mention that component.
```
```suggestion
/// Default filters are applied to any queries that does not explicitly mention that component.
```

alice-i-cecile commented on crates/bevy_ecs/src/query/default_filters.rs:
```diff
@@ -0,0 +1,173 @@
+/// component will only be visible to a [`Query`](crate::prelude::Query) containing something like
+/// [`With<Hidden>`](crate::prelude::With) or [`Has<Hidden>`](crate::prelude::Has).
+///
+/// These filters are only applied to queries initialized after updating this resource,
```
```suggestion
/// These filters are only applied to queries whose cache is generated after updating this resource.
```

alice-i-cecile commented on crates/bevy_ecs/src/query/default_filters.rs:
```diff
@@ -0,0 +1,173 @@
+/// [`With<Hidden>`](crate::prelude::With) or [`Has<Hidden>`](crate::prelude::Has).
+///
+/// These filters are only applied to queries initialized after updating this resource,
+/// it should most likely only be modified before the app starts.
```
```suggestion
/// As a result, this resource should generally only be modified before the app starts (typically during plugin construction).
```

alice-i-cecile commented on crates/bevy_app/src/app.rs:
```diff
@@ -94,6 +95,7 @@ impl Default for App {
         let mut app = App::empty();
         app.sub_apps.main.update_schedule = Some(Main.intern());
 
+        app.init_resource::<DefaultQueryFilters>();
```
This behavior should be called out in the `DefaultQueryFilters` docs.

alice-i-cecile commented on crates/bevy_ecs/src/query/default_filters.rs:
```diff
@@ -0,0 +1,173 @@
+impl DefaultQueryFilters {
+    /// Add a With filter to the default query filters.
+    /// Removes any Without filter for this component if present.
+    pub fn with(&mut self, component_id: ComponentId) {
```
These methods should be renamed to `with_untyped` etc for consistency.

Then, I think we should add methods on `World` for `default_with_filter` etc that take a type.

alice-i-cecile reviewed (changes requested):
I think this is a really solid step forward: sorry for not reviewing this earlier.

I'd like to see:

1. Mildly improved docs.
2. An example.
3. Methods to set these on `World`.
4. Minor cleanup comments addressed (either tell me I'm wrong or fix them).

alice-i-cecile commented on crates/bevy_ecs/src/world/mod.rs:
```diff
@@ -1074,6 +1075,32 @@ impl World {
 
+    /// Add a [`With<T>`](crate::prelude::With) filter to [`DefaultQueryFilters`].
+    /// Inserts the [`DefaultQueryFilters`] resource if it does not exist
+    pub fn default_with_filter<T: Component>(&mut self) {
```
I think this would be clearer as "set_default_with_filter" (and the same below).

alice-i-cecile commented on crates/bevy_ecs/src/world/mod.rs:
```diff
@@ -1074,6 +1075,32 @@ impl World {
+
+    /// Add a [`Without<T>`](crate::prelude::Without) filter to [`DefaultQueryFilters`].
+    /// Inserts the [`DefaultQueryFilters`] resource if it does not exist
+    pub fn default_without_filter<T: Component>(&mut self) {
```
We're missing a typed version of the "clear filter" API.

alice-i-cecile reviewed (approved):
Very nice: good cleanup. There's a couple of other things I noticed but they're not blocking.

I'm comfortable proceeding with this as is (pending a second SME-ECS approval). I think that a "canonical" ignore-by-default marker or three will be important, but I want to see what the community uses this for and needs first.

re0312 commented:
The code and documentation are very clean and impressive ,but current impl is unsound. The following code will panic. To make it work correctly, we should only allow setting table component filter.
```
    let mut world = World::new();
    world.spawn((Table(1), Sparse(1)));
    world.spawn(Table(2));

    world.set_default_with_filter::<Sparse>();

    let mut query = QueryState::<&Table>::new(&mut world);
    // panic!
    assert_eq!(1, query.iter(&world).count());
```
IMO ,global query configuration will greatly increase the coupling between different parts of the code. If we decide to land it, we should mark the relevant APIs as `Unsafe` .

Additionally, considering that it will likely be widely used by third-party libraries in the foreseeable future, the configuration of one plugin may easily disrupt the usage of another plugin, even if they appear unrelated.This will undoubtedly increase the maintenance cost for third parts author as they must consider compatibility with other unofficial plugins.

I personally lean towards implementing it after query as entities landed. it will unlock query-level custom settings, which could have more granular control, However, considering the strong demand for this feature within the community and the current implementation's adequacy,I'll defer my opinion to those with a more comprehensive understanding to decide.

NiseVoid commented:
> but current impl is unsound

Looking at the behavior it doesn't look like it's unsound in the usual Rust way, but this is definitely a potential footgun since it just straight up ignores the added filter without even giving you a warning.

> To make it work correctly, we should only allow setting table component filter.

This does seem like the right call, in theory we could fix the filtering for sparse components, but setting a default filter for a sparse component would be very expensive and doesn't seem like something we should allow.

@alice-i-cecile what is your opinion on this? And if we do block it, how should we let the user know, call a `warn!` and then ignore the filter?

> IMO ,global query configuration will greatly increase the coupling between different parts of the code. If we decide to land it, we should mark the relevant APIs as `Unsafe` .

I don't think this matches what should be considered unsafe in Rust. As long as no memory safety invariants are introduced the code should be safe, even if you can use it to create logic errors.

> Additionally, considering that it will likely be widely used by third-party libraries in the foreseeable future, the configuration of one plugin may easily disrupt the usage of another plugin, even if they appear unrelated.This will undoubtedly increase the maintenance cost for third parts author as they must consider compatibility with other unofficial plugins.

This shouldn't be a major problem as long as you don't use any libraries that abuse this feature, you would specifically need to have two systems adding a filtered component at the same time (between command queue flushes) to see any issues. The chance of such issues happening would of course be greatly reduced once we build a higher level disabling API on top of this feature.

SkiFire13 commented:
> it just straight up ignores the added filter without even giving you a warning.

However those filters were used to ensure that the query doesn't conflict with other system parameters, ignoring them would mean that now it can conflict with them.

I haven't actually tested this, but consider a situation like this:

- a default filter with `With<Sparse>`
- a query `Query<&mut Dense>`, which gets the default filter;
  - access-wise this is equivalent to `Query<&mut Dense, With<Sparse>>`
  - in practice it actually fetches the same entities as `Query<&mut Dense>`
- another query `Query<&mut Dense, Without<Sparse>>`
  - in theory this does not conflict with the previous query because the two filters do no overlap
  - in practice however the previous query also fetches the `&mut Dense` fetched by this query, so they do overlap, hence this is unsound

NiseVoid commented:
Oh, yes, that does indeed make it unsound. Either way we definitely need to fix it by either 1. preventing sparse filters from being added to the list; or 2. actually filtering them. Even if you can't create UB with it, the current behavior would be extremely confusing :thinking:

re0312 commented:
> This does seem like the right call, in theory we could fix the filtering for sparse components, but setting a default filter for a sparse component would be very expensive and doesn't seem like something we should allow.

It's not easy to fix sparse components without a performance drop. IIRC  to address it to support sparse compoent safely, we must introduce a branch in the hot path, which could significantly impact our performance.

> I don't think this matches what should be considered unsafe in Rust. As long as no memory safety invariants are introduced the code should be safe, even if you can use it to create logic errors.

It may not be a memory issue, but the impact is too extensive and could greatly increase the vulnerability of the code. I strongly recommend marking it as unsafe so that people who use it should fully understand the implications of their action.

Additionally, `set_default_with_filter `only affects queries created before it's set. As we widely use cached queries in the system, it does not affect cached-queries after schedule initialization, even if we set it later. This means if we set it before system initialization, the influence will be global and cannot be canceled at runtime. If we set it at runtime to build a query, why not use QueryBuilder to create what we want instead?

```

systemA(world:&mut World){
    world.set_default_with_filter::<A>();   
}

SystemB(query:Query<&B>){

    // The system query will not be affected by set_default_with_filter 
    // because its query state was created during schedule initialization.
    --------------------
}

App.add_systems(Update ,(SystemA,SystemB).chain());
```

> This shouldn't be a major problem as long as you don't use any libraries that abuse this feature, you would specifically need to have two systems adding a filtered component at the same time (between command queue flushes) to see any issues. The chance of such issues happening would of course be greatly reduced once we build a higher level disabling API on top of this feature.

SkiFire13 commented:
Note that while restricting filters to dense components would fix this issue, it would prevent future optimizations to sparse only queries (e.g. `Query<&Sparse>` could be optimized to iterate over only the relative sparse set instead of iterating the various archetypes first). This is a tradeoff between what you can statically assume at compile time for optimizations and what you want to make dynamic.

iiYese commented:
From exploring this as the only viable alternative to entity disabling we've discovered:
- Unsoundness from the limitations of current ECS internals
- [Flaws with the approach before in #ecs-dev](https://discord.com/channels/691052431525675048/749335865876021248/1250856257238204497)

I think this should be enough conclusive motivation to push forward with the original entity disabling PR #12928 which was simpler and didn't have these problems. Most of the issues brought up with entity disabling were to do with poorly understood/conveyed semantics [which is mostly down to naming](https://discord.com/channels/691052431525675048/749335865876021248/1250921626023755856) and an issue that should be solved with docs rather than a completely different feature.

@NiseVoid @alice-i-cecile @JoJoJet

NiseVoid commented:
If we limit the components you can register to ones with table storage, the unsoundness would be solved, and the only extra issue this approach introduces is the ordering (adding filters after queries are built), which can be solved once we have some way to access query caches and rebuild them (like Queries as Entities). The other mentioned issues (spare optimization and sander's comment on discord) already exist on the `Disabled` PR.

I think comparing the PRs as different solutions is weird, the `Disabled` PR relies on this behavior, it's just hardcoded into `QueryState` there, this PR pulls the behavior out into a reusable form. The goal being to allow third-party crates to experiment rather than trying to get first-party entity disabling merged entirely based off speculation.

SkiFire13 commented:
> If we limit the components you can register to ones with table storage, the unsoundness would be solved

Note that this fix still leaves the possibility for future unsoundness if any new optimizations are done on `WorldQuery`. This essentially ties the `unsafe` internal optimizations of `WorldQuery` with default query filters.

> and the only extra issue this approach introduces is the ordering (adding filters after queries are built), which can be solved once we have some way to access query caches and rebuild them (like Queries as Entities)

Will this work with manual `QueryState`/`SystemState`? For example if are stored in a `Local`?

NiseVoid commented:
> Note that this fix still leaves the possibility for future unsoundness if any new optimizations are done on `WorldQuery`. This essentially ties the `unsafe` internal optimizations of `WorldQuery` with default query filters.

I think those optimizations would all touch `QueryState`, and these filters are also applied there. This unsound behavior could also probably be caused trough other means, since it's ultimately caused by some odd behavior in QueryState and the coupling of filters and the rules used to avoid conflicting queries.

> Will this work with manual `QueryState`/`SystemState`? For example if are stored in a `Local`?

I think that ultimately depends on the implementation, but considering updating caches is necessary for a few prerequisites of relations I'd assume that will work just fine, unless it's a non-cached query, in which case we could just have it always fetch the current filters

JeanMertz commented:
> I think this should be enough conclusive motivation to push forward with the original entity disabling PR #12928 which was simpler and didn't have these problems.

I would note that PR triggered the creation of a (draft) RFC at https://github.com/bevyengine/rfcs/pull/81. Given the back-and-forth discussion and open questions by @mockersf in https://github.com/bevyengine/bevy/pull/12928#issuecomment-2054175843, it does seem desirable to push that RFC forward first, get it signed off, and _then_ get #12928 into a mergeable state.

NiseVoid commented:
I think with @SkiFire13's fix (#14615) it should be possible to fix the sparse-related problems. I have updated the PR and added a test for it and it seems to work as one would expect. I have somehow never been able to reproduce @re0312's panic, but it also completely ignored sparse filters previously, which caused unsound behavior as queries that got detected as not overlapping could alias mutable references.

alice-i-cecile commented:
After [discussion with @maniwani](https://discordapp.com/channels/691052431525675048/749335865876021248/1286035837305229342), we have approval from two SME-ECS: them and myself.

As a follow-up, we should add a first-party `Disabled` marker component for ecosystem coordination purposes, and include a default default query filter.

iiYese commented:
I'm going to raise a concern I've already mentioned on discord. If we have a `Disabled` marker it is easy to define what is considered erroneous for that. Eg. mutating an entity while it's disabled. It's also easy for authors to handle the addition & removal of `Disabled` via hooks & observers. This is not the case for default filters. Default filters will have completely different semantics depending on the author & there is no way for authors to handle the addition & removal of all filters from other authors. It would create users needing to request plugin authors for compatibility hooks for other plugins.

alice-i-cecile commented:
@NiseVoid I know you've been chewing on designs for this, what's the plan for this work? My current understanding is:

1. Ship something shaped broadly like this PR, but make it fully pub(crate)
2. Add a trivial `Disabled` component, and add it to `DefaultQueryFilters` via a plugin.
3. Figure out how to make it play nicer with hierarchies.
4. Swap `OnAdd` and friends to trigger when entities are disabled.

Is that correct? What needs to be done to this PR to get step 1 above mergeable?

NiseVoid commented:
> @NiseVoid I know you've been chewing on designs for this, what's the plan for this work? My current understanding is:
> 
> 1. Ship something shaped broadly like this PR, but make it fully pub(crate)
> 2. Add a trivial `Disabled` component, and add it to `DefaultQueryFilters` via a plugin.
> 3. Figure out how to make it play nicer with hierarchies.
> 4. Swap `OnAdd` and friends to trigger when entities are disabled.
>
> Is that correct? What needs to be done to this PR to get step 1 above mergeable?

Yes, this is correct. For docs and lint reasons making it `pub(crate)` in this PR might be difficult (doc tests would fail to compile and all of the code would get marked dead), but I have some changes that should generally prevent the issues @iiYese mentioned. Mainly you can only set one filter "per concept" (so entity disabling gets 1 filter, if we later add prefabs we could add one there too). Once we add `Disabled`, if we're okay with that living in `bevy_ecs` we can make the setter for the disabled component id `pub(crate)`, and only expose a way to `init_component::<Disabled>()` and set that `ComponentId` as the filter.

alice-i-cecile commented:
Sounds good. Feel free to add expect(dead_code) to get CI to pass for now :)

NiseVoid commented:
@alice-i-cecile Where do you think docs for entity disabling and other similar features should live? Currently there's some docs and a doc test on `DefaultQueryFilters`, and with the limited scope of this PR I can't currently both have doc test and make `set_disabled` `pub(crate)` since doc tests can only access public functions. If we don't want `DefaultQueryFilters` to be the primary place explaining the features removing the doc test for the now simplified version wouldn't be a huge deal, but if it is the right place going forward we might need to merge the PR with that function public and make it not public and replace it with something less abusable like a function that calls `set_disabled` with the `ComponentId` for `Disabled`.

Also comments on the last commit would be welcome, I simplified the design somewhat for the more locked down design, but some parts of it do feel a bit odd like the `Option<()>` on `set_disabled` or working on an iterator that isn't currently necessary (but should simplify extending it with say `Prefab` or `InternalMarkerPlsNoTouch` later)

alice-i-cecile commented:
Can we make an `entity_disabling` module that stores these types? And then `Disabled` can live in there as one of the few `pub` types?

alice-i-cecile reviewed (approved):
@maniwani  and I have signed off on the core design as SME-ECS (hence the change to X-Blessed a while back). The core challenge here is how to handle disabling entities WRT relations. After discussing this with Sander Martens (of flecs), I'm of the opinion that the simple, intuitive behavior here is the right one to ship with, while making sure that it's easy to "disable/enable recursively".

Before 0.16, I think it's important that we:

1. Ship a first party `Disabled` component.
2. Demonstrate this feature in an example, which showcases the hierarchical disabling pattern.

@NiseVoid has suffered long enough waiting for this to land though, and those are easily moved to follow-up. Merging.

alice-i-cecile commented:
Thank you to everyone involved with the authoring or reviewing of this PR! This work is relatively important and needs release notes! Head over to https://github.com/bevyengine/bevy-website/issues/1982 if you'd like to help out.
